File: 15. 3Sum/HashSetSortingSolution.java
import java.util.*;

class HashSetSortingSolution 
{
    public List<List<Integer>> threeSum(int[] nums) 
    {
        
        int arrayLength = nums.length;

		//Initializing the target value for this generic 3-sum problem.
        int target = 0;

		//Creating a result HashSet of Lists to avoid duplicate entries
        HashSet<List<Integer>> resultset = new HashSet<>();

		//Iterating through the array
        for (int i = 0; i < arrayLength; i++)
        {
	
			//Creating a HashSet to store the nums[j] values traversed so far
            HashSet<Integer> visitedSet = new HashSet<>();

            for (int j = i+1; j < arrayLength; j++)
            {
				//Calculating the complement and checking if it exists inside the set of elements that have been traversed from i+1 to the current jth-1 value
                int complement = target - nums[i] - nums[j];

				//If the value is previously seen then we have found our combination
                if(visitedSet.contains(complement))
                {
					//Creating a List of the result values
                    List<Integer> result = Arrays.asList(nums[i],nums[j],complement);

					//Sorting these values so that -1,2,-1 and 2,-1,-1 for example are considered as the same solution.
                    Collections.sort(result);

					//Adding this combination to our result set
                    resultset.add(result);
                }

				//Adding this current nums[j] value to our set as it is already visited.
                visitedSet.add(nums[j]);
            }
        }

		//Converting the result set to an ArrayList as it is the expected return type.
        return new ArrayList<>(resultset);
    }
}

File: 15. 3Sum/HashSetSortingSolution.py
class HashSetSortingSolution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        #This below code is to handle the edge case of all numbers being 0 in the list. Instead of checking combinations for all of them we can simply return the result [[0,0,0]] as the answer will be the same.

        count0 = nums.count(0)
        if count0 == len(nums):
            return [[0,0,0]]

        #Code for the rest of the cases

        n = len(nums)

        #This is a generic 3sum problem where target can be any number, here it is 0
        target = 0
	
        #Using a set to store the combinations to take care of duplicate values
        resultSet = set()
        
        # Generating all pairs
        for i in range(n):

            #Creating a set for each iteration of i to keep a track of every j'th value that we have seen so far.
            visitedSet = set()

            for j in range(i + 1, n):

                #Calculating the complement that we need to look for in the set.
                complement = target - nums[i] - nums[j]
                
                #Check if this complement is already an element if we have seen before, if yes then we have found our triplet
                if complement in visitedSet:

                    #Converting each solution to a tuple as it is Hashable and when we try to insert the same solution again in the result set, it will not get inserted. We sort it so that e.g. results like -1,2,-1 and 2,-1,-1 are not considered as different results by the code but same instead.
                    curr = tuple(sorted([nums[i], nums[j], complement]))
                    resultSet.add(curr)
                
                #Add the j'th element to the set of elements we have seen so far in the i'th iteration.    
                visitedSet.add(nums[j])
        
        return list(resultSet)